library/src/main/java/com/yausername/youtubedl_android/YoutubeDL.kt

```diff
@@ -16,84 +46,184 @@ import kotlin.collections.set
 
 object YoutubeDL {
     private var initialized = false
+    internal lateinit var binariesDirectory: File
+
     private var pythonPath: File? = null
     private var ffmpegPath: File? = null
-    private var ytdlpPath: File? = null
-    private var binDir: File? = null
-    private var ENV_LD_LIBRARY_PATH: String? = null
-    private var ENV_SSL_CERT_FILE: String? = null
-    private var ENV_PYTHONHOME: String? = null
-    private val idProcessMap = Collections.synchronizedMap(HashMap<String, Process>())
+    private lateinit var ytdlpPath: File
+
+    /* ENVIRONMENT VARIABLES */
+    private lateinit var ENV_LD_LIBRARY_PATH: String
+    private lateinit var ENV_SSL_CERT_FILE: String
+    private lateinit var ENV_PYTHONHOME: String
 
+    private val pluginsDownloader: PluginsDownloader = PluginsDownloaderImpl()
+
+    //Map of process id associated with the process
+    private val idProcessMap = Collections.synchronizedMap(HashMap<String, Process>())
     @Synchronized
-    @Throws(YoutubeDLException::class)
-    fun init(appContext: Context) {
+    @Throws(YoutubeDLException::class, MissingPlugin::class)
+    /**
+     * Initializes the library. This method should be called before any other method.
+     * @param appContext the application context
+     */
+    fun init(appContext: Context, callback: pluginDownloadCallback? = null) {
```

Comment:
I prefer not to modify the previous APIs, especially this one, which might bring too many breaking changes.

Reply:
> We should maintain the original method of loading dependencies from the APK or external files, since network connections are not always reliable or simply for compatibility.

We can simply implement extract their common parts, and then separate the inconsistent logics/files into two methods or even two libraries

Reply:
I have moved it to another method. In case that you want to move it to another library tell me; we have to modify both the FFMpeg module and Aria2c.
I was thinking on merging them into just the `library` module, so we can share the code. In other case we would have to make the thing you said before, making a separated library, but that would be very much job.

Reply:
Imagine the case that we need to build two variations of the same application, one using bundled-in dependencies and the other download them via internet connections. (Yes we do want to achieve this as I've mentioned before)

Spliting libraries solution: making the main `library` code able to handle these two cases, then package the native dependencies into standalone libraries. So we'll able to load the dependencies by simply include them in the gradle build scripts, or download them from specific URLs when the dependency libraries absent

Reply:
Okay, so if I understood well the idea is to make the library be able to build in two different ways; one with the libraries already bundled, and another one without them, no?